features/util.py

```diff
@@ -161,3 +161,33 @@ def lxc_get_series(name: str, image: bool = False):
                 " Could not detect image series. Add it via `lxc image edit`"
             )
     return None
+
+
+def lxc_push_source_pull_deb_pkg(name: str) -> None:
+    """
+    Push source PR code .tar.gz to the container.
+    Run tools/build-from-source.sh which will create the .deb
+    Pull .deb from this container to travis-ci instance
+
+    :param name:
+        name of the container to:
+         - push the PR source code;
+         - pull the built .deb package.
+
+    """
+    print ('\n\n\n LXC file push pr_source.tar.gz')
+    subprocess.run(["lxc", "file", "push", "/tmp/pr_source.tar.gz", name+'/tmp/'])
+    print ("\n\n Extract /tmp/pr_source.tar.gz")
+    subprocess.run(["lxc", "exec", name, "--", "tar", "-xzvf", "/tmp/pr_source.tar.gz", "--directory", "/tmp/"])
+    print ('\n\n\n Run /tools/build-from-source.sh')
+    subprocess.run(
+        [
+            "lxc",
+            "exec",
+            name,
+            "--",
+            "/tmp/ubuntu-advantage-client/tools/build-from-source.sh"
+        ],
+    )
+    print ("\n\nPull .deb from the instance to travis VM")
+    subprocess.run(["lxc", "file", "pull", name+'/tmp/ubuntu-advantage-tools_20.4_amd64.deb', "/tmp/"])
```

Comment:
1. I think we can avoid having to add a file in  ./tools/build-from-source.sh  if we just do this inside behave.
2. we should use the lxc_exec function wherever we can instead lf subprocess.run directly
3. it might be worth adding an lxc_file_push function

```suggestion

    # create the script file
    with open('build-from-source.sh', 'w') as stream
        stream.write(textwrap.dedent(""\
            #!/bin/bash
            set -o xtrace
            apt-get update
            apt-get install make
            cd /tmp
            tar -zxvf *gz
            cd ubuntu-advantage-client
            make deps
            dpkg-buildpackage -us -uc
         """))
      
    lxc_file_push("build-from-source.sh", container_name + "/tmp/")
    print ('\n\n\n Run /tools/build-from-source.sh')
    lxc_exec(container_name, ["--", "bash", "/tmp/build-from-source.sh"])
    print ("\n\nPull .deb from the instance to travis VM")
    # rename the output deb to ubuntu-advantage.deb so we don't have to worry about version changes.
    subprocess.run(["lxc", "file", "pull", container_name+'/tmp/ubuntu-advantage-tools*.deb', output_deb_file])
```

Reply:
@blackboxsw I cannot use the `*` here, according to this issue: https://github.com/lxc/lxd/issues/3523
it is not supported

Reply:
I changed its name inside the new generated script, so we do not need to worry about version changes

Reply:
Also, as lxc_exec is using --user 1000  we cannot run `lxc_exec(container_name, ["--", "bash", "/tmp/build-from-source.sh"])` or will get an error
```
+ apt-get update                                                
E: Could not open lock file /var/lib/apt/lists/lock - open (13: Permission denied)
E: Unable to lock directory /var/lib/apt/lists/                      
E: Could not open lock file /var/lib/dpkg/lock - open (13: Permission denied)
```
E: Unable to lock the administration directory (/var/lib/dpkg/), are you root?```
but just adding sudo does work, so moving to it, sorry.